Add WKWebsiteDataTypeHashSalt to allow fetching/removing DeviceIDHashSalt data

szewai opened the pull request:
#### 2630f203a0c9c9bb0ff5e884d47254a6a8101e76
<pre>
Add WKWebsiteDataTypeHashSalt to allow fetching/removing DeviceIDHashSalt data
<a href="https://bugs.webkit.org/show_bug.cgi?id=253174">https://bugs.webkit.org/show_bug.cgi?id=253174</a>
rdar://106096684

Reviewed by Chris Dumez.

Otherwise the data type cannot be viewed by WebKit client.

* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecord.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecord.mm:
(dataTypesToString):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordInternal.h:
(WebKit::toWebsiteDataType):
(WebKit::toWKWebsiteDataTypes):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(+[WKWebsiteDataStore allWebsiteDataTypes]):

Canonical link: <a href="https://commits.webkit.org/261194@main">https://commits.webkit.org/261194@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ae8f5d72ec641f46a63c568ce8617cbcf4d7bf5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110534 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19616 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43116 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1898 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119414 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21046 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10758 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/1366 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116276 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15680 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98863 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102767 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97640 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30529 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43928 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12271 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31867 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85797 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12843 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8836 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18207 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51484 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7755 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14714 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/workers/service/server/RegistrationDatabase.cpp:
```diff
@@ -140,6 +140,7 @@ RegistrationDatabase::RegistrationDatabase(RegistrationStore& store, String&& da
     , m_databaseDirectory(WTFMove(databaseDirectory))
     , m_databaseFilePath(FileSystem::pathByAppendingComponent(m_databaseDirectory, databaseFilename()))
 {
+    WTFLogAlways("sihuil:[%p]RegistrationDatabase::RegistrationDatabase m_databaseFilePath[%s]", this, m_databaseFilePath.utf8().data());
```
not needed

szewai replied:
Ah will removed.

youennf commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -687,7 +687,7 @@ void WebsiteDataStore::removeData(OptionSet<WebsiteDataType> dataTypes, WallTime
     }
 
-    if (dataTypes.contains(WebsiteDataType::DeviceIdHashSalt) || (dataTypes.contains(WebsiteDataType::Cookies)))
+    if (dataTypes.contains(WebsiteDataType::DeviceIdHashSalt))
```
I am not sure we should do this change.
Spec states: `[User Agents](https://w3c.github.io/mediacapture-main/#dfn-user-agent) MUST rotate per-origin device identifiers when other persistent storage are cleared.`

szewai replied:
ah so this suggests device id should have the same lifetime as cookies. In this case do we think HashSalt is cookies?
If it is:
Then we should return WKWebsiteDataTypeCookies for website data fetch when clients have DeviceIdHashSalt data, and remove both cookies and hash salt when clients remove WKWebsiteDataTypeCookies data.

If it is not:
Then we should return WKWebsiteDataTypeHashSalt for website data fetch, and let client remove data for WKWebsiteDataTypeHashSalt. (We may keep the existing behavior that removes WKWebsiteDataTypeHashSalt data when clients remove WKWebsiteDataTypeCookies)

Which one do you think it should be?

youennf replied:
I would do the latter, expose it but continue removing it when cookies are cleared.

szewai replied:
Ok, the patch is updated as suggested

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecord.h:
```diff
@@ -68,6 +68,9 @@ WK_EXTERN NSString * const WKWebsiteDataTypeSearchFieldRecentSearches WK_API_AVA
 /*! @constant WKWebsiteDataTypeMediaKeys MediaKeys storage */
 WK_EXTERN NSString * const WKWebsiteDataTypeMediaKeys WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
 
+/*! @constant WKWebsiteDataTypeHashSalt Hash salt for deviceId */
```
Ok, though after reading this comment I still have no idea what this clears.

szewai replied:
It's storage used for this type: https://developer.mozilla.org/en-US/docs/Web/API/MediaDeviceInfo/deviceId...

szewai committed:
Add WKWebsiteDataTypeHashSalt to allow fetching/removing DeviceIDHashSalt data
https://bugs.webkit.org/show_bug.cgi?id=253174
rdar://106096684

Reviewed by Chris Dumez.

Otherwise the data type cannot be viewed by WebKit client.

* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecord.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecord.mm:
(dataTypesToString):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordInternal.h:
(WebKit::toWebsiteDataType):
(WebKit::toWKWebsiteDataTypes):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(+[WKWebsiteDataStore allWebsiteDataTypes]):

Canonical link: https://commits.webkit.org/261194@main
